src/drivers/usb/class/usb_class_cdc.c

```diff
@@ -12,100 +14,244 @@
 #include <kernel/panic.h>
 #include <embox/unit.h>
 #include <drivers/usb/usb_cdc.h>
+#include <drivers/usb/usb_driver.h>
 #include <kernel/printk.h>
 #include <stdlib.h>
+#include <string.h>
 
 EMBOX_UNIT_INIT(usb_cdc_init);
 
 POOL_DEF(cdc_classes, struct usb_class_cdc, USB_CDC_MAX_DEVS);
 
-#if 0
-static uint32_t line_coding;
+/* Storage for device configuration*/
+static char cdc_conf[16];
 
-static void get_line_coding_conf_hnd(struct usb_request *req, void *arg) {
+#define INTERFACE_DESC_OFFSET          (sizeof (struct usb_desc_configuration))
+#define CS_INTERFACE                   0x24
+
+/* Configuration */
+static void cdc_get_conf_len_hnd(struct usb_request *req, void *arg);
+static void cdc_get_conf(struct usb_dev *dev, struct usb_class_cdc *cdc,
+		size_t idx, size_t len, usb_request_notify_hnd_t cb);
+static void cdc_get_conf_content_hnd(struct usb_request *req, void *arg);
+static void cdc_get_conf_hnd(struct usb_request *req, void *arg);
+static void cdc_next_conf(struct usb_dev *dev);
+
+/* Endpoint */
+static size_t cdc_skip_func_descs(char *start);
+static size_t cdc_skip_endpoints(char *start);
+static void cdc_get_endpoints(struct usb_dev *dev, struct usb_class_cdc *cdc);
+
+/* Misc */
+static int is_rndis(struct usb_desc_interface *desc);
+
+/**
+ * Skip all functional descriptors.
+ * Return size of all functional descriptors.
+ * @see ecm120.pdf (Subclass Specification for
+ *  Ethernet Control Model Devices) Section 5
+ * XXX skip is a default action now, but it is possible to redefine this behavior
+ */
+static size_t cdc_skip_func_descs(char *start) {
+	char *cur = start;
+
+	while (1) {
+		uint8_t len, type;
+
+		type = *(cur + 1);
+		if (type == CS_INTERFACE) {
+			len = *cur;
+			cur += len;
+		} else {
+			break;
+		}
+	}
+
+	return (cur - start);
 }
 
-static void get_line_coding(struct usb_dev *dev) {
-	line_coding = 18;
+/*
+ * @see ecm120.pdf (Subclass Specification for
+ *  Ethernet Control Model Devices) Section 5
+ */
+static size_t cdc_skip_endpoints(char *start) {
+	char *cur = start;
 
-	usb_endp_control(dev->endpoints[0], get_line_coding_conf_hnd, NULL,
-			0x21,
-			0x20,
-			0,
-		dev->c_config,
-		sizeof line_coding,
-		&line_coding);
+	while (1) {
+		uint8_t len, type;
 
-	line_coding = 0;
+		type = *(cur + 1);
+		if (type == USB_DESC_TYPE_ENDPOINT) {
+			len = *cur;
+			cur += len;
+		} else {
+			break;
+		}
+	}
 
-	usb_endp_control(dev->endpoints[0], get_line_coding_conf_hnd, NULL,
-			0xA1,
-			0x21,
-			0,
-		dev->c_config,
-		sizeof line_coding,
-		&line_coding);
+	return (cur - start);
+}
 
-	printk("line_coding - %d\n", line_coding);
+struct usb_desc_interface *cdc_get_interface(
+		struct usb_desc_configuration *conf, size_t index) {
+	size_t cur = 0;
+	char *curp = (char *) conf + INTERFACE_DESC_OFFSET;
+
+	while (cur != index) {
+		curp += sizeof (struct usb_desc_interface);
+		curp += cdc_skip_func_descs(curp);
+		curp += cdc_skip_endpoints(curp);
```

Comment:
Yes. Endpoints always go after interface, while func descs are part of interface.